### PR TITLE
WIP : Code changes for BundleExport

### PR DESCRIPTION
**Pull request by vinu2003**
This PR has server side code changes for export bundle feature.

## QA steps
execute juju export-bundle command in developer-mode
## Documentation changes
N/A
## Bug reference
N/A

yaml file generated by the command export-bundle for mediawiki single:  https://pastebin.ubuntu.com/p/Kg38k6KCc5/

### CODE REVIEW

**Comment by vinu2003** [comment]
Hi @howbazaar , I have verified the bundle export cmdline functionality by merging tthis PR with PR#8912 and deployed a simple 'media wiki single' bundle from the .yaml file generated using GUI and from command line exported the model config and used the bundle file generated by the juju export-bundle command to deploy back successfully.

File: apiserver/facades/client/bundle/bundle.go [hunk 1]
```diff
@@ -151,10 +156,111 @@ func (b *BundleAPI) GetChanges(args params.BundleChangesParams) (params.BundleCh
...
+	return params.BytesResult{
+		Result: bytes,
+	}, nil
 }
 
 // Mask the new method from V1 API.
 // ExportBundle is not in V1 API.
 func (u *APIv1) ExportBundle() (_, _ struct{}) { return }
+
+// Fills the bundledata datastructure required for the exportBundle.
```
Comment (anastasiamac) [inline apiserver/facades/client/bundle/bundle.go]: Should match method name, i.e.

```
// FillBundleData ...
```

File: apiserver/facades/client/bundle/bundle.go [hunk 2]
```diff
@@ -151,10 +156,111 @@ func (b *BundleAPI) GetChanges(args params.BundleChangesParams) (params.BundleCh
...
+func (b *BundleAPI) FillBundleData(model description.Model) (*charm.BundleData, error) {
```
Comment (anastasiamac) [inline apiserver/facades/client/bundle/bundle.go]: Any reason why this is an exported method?

It would be helpful if we had an interface declaration for v1 and v2 that would clearly highlight the differences between versions. This would also allow us to track differences in the future when we'd be at 3, 4, 5, etc

File: apiserver/facades/client/bundle/bundle.go [hunk 3]
```diff
@@ -151,10 +156,111 @@ func (b *BundleAPI) GetChanges(args params.BundleChangesParams) (params.BundleCh
 }
 
 // ExportBundle exports the current model configuration as bundle.
-func (b *BundleAPI) ExportBundle() (params.StringResult, error) {
-	return params.StringResult{}, nil
+func (b *BundleAPI) ExportBundle() (params.BytesResult, error) {
```
Comment (anastasiamac) [inline apiserver/facades/client/bundle/bundle.go]: We cannot just remove 
```
ExportBundle() (params.StringResult, error)
```
as v1 should still have it to comply with the original promise.

Also why would we return a BytesResult? or even a StringResult? We customary return map[string]interface{} for similar situations and let the clients parse the results. For example, see modelmanager.DumpAll()

File: apiserver/facades/client/bundle/bundle.go [hunk 4]
```diff
@@ -151,10 +156,111 @@ func (b *BundleAPI) GetChanges(args params.BundleChangesParams) (params.BundleCh
...
 }
 
 // Mask the new method from V1 API.
 // ExportBundle is not in V1 API.
 func (u *APIv1) ExportBundle() (_, _ struct{}) { return }
+
+// Fills the bundledata datastructure required for the exportBundle.
+func (b *BundleAPI) FillBundleData(model description.Model) (*charm.BundleData, error) {
+	data := &charm.BundleData{
+		Series:       series.LatestLts(),
```
Comment (anastasiamac) [inline apiserver/facades/client/bundle/bundle.go]: I am not sure that we want to hard-code it to be latest LTS. Maybe, it should be a juju supported latest lts? However, I would have expected there to be some logic calculation, maybe based on the series of the applications that are in this bundle once we've obtained them?...
Is this what javascript client currently does? We do want the parity, and if it is, this place deserves at least a comment to this effect with a TODO to address.

File: apiserver/facades/client/bundle/bundle.go [hunk 5]
```diff
@@ -151,10 +156,111 @@ func (b *BundleAPI) GetChanges(args params.BundleChangesParams) (params.BundleCh
...
+			ut = append(ut, unit.Machine().Id())
+		}
+
+		newApplication := &charm.ApplicationSpec{
+			Charm:            application.CharmURL(),
+			Series:           application.Series(),
+			NumUnits:         len(application.Units()),
+			To:               ut,
+			Expose:           application.Exposed(),
+			Options:          application.CharmConfig(),
```
Comment (anastasiamac) [inline apiserver/facades/client/bundle/bundle.go]: This does not look right.
CharmConfig() are the settings that charm declares; ApplicationConfig() are the settings that application uses.
Options here need to be application config.

File: apiserver/facades/client/bundle/bundle.go [hunk 6]
```diff
@@ -151,10 +156,111 @@ func (b *BundleAPI) GetChanges(args params.BundleChangesParams) (params.BundleCh
...
+			Annotations:      application.Annotations(),
+			EndpointBindings: application.EndpointBindings(),
+		}
+
+		data.Applications[application.Name()] = newApplication
+	}
+
+	for _, machine := range model.Machines() {
+		constraints := []string{"arch=" + machine.Constraints().Architecture(),
```
Comment (anastasiamac) [inline apiserver/facades/client/bundle/bundle.go]: What happens when machine.Constraints() is empty? It is defined as "omitempty" after all.
Also all individual properties are defined in the same way, so they can be empty too and need to be catered for...

File: apiserver/facades/client/bundle/bundle_test.go [hunk 7]
```diff
@@ -214,3 +215,14 @@ func (s *bundleSuite) TestGetChangesBundleEndpointBindingsSuccess(c *gc.C) {
...
 }
+
+func (s *bundleSuite) TestExportBundleFailNoApplication(c *gc.C) {
+	var cfg state.ExportConfig
+	model, err := s.st.ExportpartialNoApplication(cfg)
+	c.Assert(err, jc.ErrorIsNil)
+
+	c.Assert(model.Validate(), jc.ErrorIsNil)
+
+	_, err = s.facade.FillBundleData(model)
```
Comment (anastasiamac) [inline apiserver/facades/client/bundle/bundle_test.go]: Shouldn't this be calling .ExportBundle() instead?

File: apiserver/facades/client/bundle/bundle.go [hunk 8]
```diff
@@ -151,10 +156,111 @@ func (b *BundleAPI) GetChanges(args params.BundleChangesParams) (params.BundleCh
...
+func (b *BundleAPI) FillBundleData(model description.Model) (*charm.BundleData, error) {
+	data := &charm.BundleData{
+		Series:       series.LatestLts(),
+		Applications: make(map[string]*charm.ApplicationSpec),
+		Relations:    [][]string{},
+		Machines:     make(map[string]*charm.MachineSpec),
+	}
+
+	if len(model.Applications()) == 0 {
+		return &charm.BundleData{}, errors.NotFoundf("application")
```
Comment (anastasiamac) [inline apiserver/facades/client/bundle/bundle.go]: Is this going to make sense to a user? They have not requested a particular application.

I think we want to say here that there was nothing to export because this model had no applications... or similar message.

File: apiserver/facades/client/bundle/mock_test.go [hunk 9]
```diff
@@ -5,10 +5,48 @@ package bundle_test
 
 import (
 	"github.com/juju/testing"
+	"gopkg.in/juju/names.v2"
+
+	"github.com/juju/description"
```
Comment (anastasiamac) [inline apiserver/facades/client/bundle/mock_test.go]: move me

File: apiserver/facades/client/bundle/mock_test.go [hunk 10]
```diff
@@ -5,10 +5,48 @@ package bundle_test
...
+	"github.com/juju/juju/state"
 )
 
 type mockState struct {
 	testing.Stub
+	bundle.Backend
+}
+
+func (m *mockState) SetExportconfig(cfg state.ExportConfig) {
+	cfg.SkipActions = true
```
Comment (anastasiamac) [inline apiserver/facades/client/bundle/mock_test.go]: This would be redundant, especially if this config never changes for bundle export.

File: apiserver/facades/client/bundle/mock_test.go [hunk 11]
```diff
@@ -5,10 +5,48 @@ package bundle_test
...
+	cfg.SkipCloudImageMetadata = true
+	cfg.SkipCredentials = true
+	cfg.SkipIPAddresses = true
+	cfg.SkipSSHHostKeys = true
+	cfg.SkipStatusHistory = true
+	cfg.SkipLinkLayerDevices = true
+}
+
+func (m *mockState) ExportpartialNoApplication(config state.ExportConfig) (description.Model, error) {
```
Comment (anastasiamac) [inline apiserver/facades/client/bundle/mock_test.go]: This does not look right.

For one mock state can only implement the methods from an actual state.
There are couple of other things here. Let's talk through them f-2-f.

File: apiserver/facades/client/bundle/state.go [hunk 12]
```diff
@@ -3,4 +3,15 @@
...
 package bundle
 
-type Backend interface{}
+import (
+	"github.com/juju/juju/migration"
+	"github.com/juju/juju/state"
+)
+
+type Backend interface {
+	migration.StateExporter
```
Comment (anastasiamac) [inline apiserver/facades/client/bundle/state.go]: I am not sure that we actually want this interface here (nor this import)... Nor do I think you should change migration interface for the purpose of the bundle export!
 I think it should be just a method declaration  you are after...

```
ExportPartial(cfg state.ExportConfig) (description.Model, error)
```

File: migration/migration.go [hunk 13]
```diff
@@ -29,6 +29,7 @@ var logger = loggo.GetLogger("juju.migration")
 type StateExporter interface {
 	// Export generates an abstract representation of a model.
 	Export() (description.Model, error)
+	ExportPartial(cfg state.ExportConfig) (description.Model, error)
```
Comment (anastasiamac) [inline migration/migration.go]: This does not sound valid as we do not want to affect anything to do with migrations :D

File: apiserver/facades/client/bundle/bundle.go (hunk 3 above)
Comment (howbazaar) [reply apiserver/facades/client/bundle/bundle.go]: @anastasiamac I asked @vinu2003 to do it this way as this is the new correct way to implement versioned APIs. 

I agree that it should be a `StringResult` rather than a `BytesResult` as it will be easier for the gui to consume.

@anastasiamac for this call we are wanting to return the bundle, which is a YAML document, not a map[string]interface{}.

File: apiserver/facades/client/bundle/bundle.go [hunk 14]
```diff
@@ -151,10 +156,111 @@ func (b *BundleAPI) GetChanges(args params.BundleChangesParams) (params.BundleCh
...
 
 // ExportBundle exports the current model configuration as bundle.
-func (b *BundleAPI) ExportBundle() (params.StringResult, error) {
-	return params.StringResult{}, nil
+func (b *BundleAPI) ExportBundle() (params.BytesResult, error) {
+	if err := b.checkCanRead(); err != nil {
+		return params.BytesResult{}, common.ServerError(common.ErrPerm)
+	}
+
+	var exportConfig state.ExportConfig
```
[the remark this answers is not in the document]
Comment (howbazaar) [reply apiserver/facades/client/bundle/bundle.go]: To simplify the extraction of the model information, we are taking advantage of the ExportPartial method on state. Since this needs the ExportConfig, we are including state here for expediency.

Perhaps in the future we may add an ExportForBundle method on state if we really want to not import state here, but until we know more we aren't doing that.

File: apiserver/facades/client/bundle/bundle.go [hunk 15]
```diff
@@ -151,10 +156,111 @@ func (b *BundleAPI) GetChanges(args params.BundleChangesParams) (params.BundleCh
...
+	exportConfig.SkipSSHHostKeys = true
+	exportConfig.SkipStatusHistory = true
+	exportConfig.SkipLinkLayerDevices = true
+
+	model, err := b.backend.ExportPartial(exportConfig)
+	if err != nil {
+		return params.BytesResult{}, common.ServerError(err)
+	}
+
+	err = model.Validate()
```
Comment (howbazaar) [inline apiserver/facades/client/bundle/bundle.go]: Don't call Validate. It won't be valid. The ExportPartial is not defined to produce a valid model export.

File: apiserver/facades/client/bundle/bundle.go (hunk 2 above)
Comment (howbazaar) [reply apiserver/facades/client/bundle/bundle.go]: I agree with @anastasiamac here, the method shouldn't be exported.

File: apiserver/facades/client/bundle/bundle.go (hunk 4 above)
Comment (howbazaar) [reply apiserver/facades/client/bundle/bundle.go]: We definitely don't want to do this.

The Series for the bundle should be the default-series from the model config.

File: apiserver/facades/client/bundle/bundle.go (hunk 5 above)
Comment (howbazaar) [reply apiserver/facades/client/bundle/bundle.go]: @anastasiamac unfortunately not. Bundles don't yest support application config, that is coming later (for CAAS stuff).

The options her are charm config, but we only want to output the values that have been explicitly set, which I'm pretty sure the model export handles.

File: apiserver/facades/client/bundle/bundle.go [hunk 16]
```diff
@@ -151,10 +156,111 @@ func (b *BundleAPI) GetChanges(args params.BundleChangesParams) (params.BundleCh
...
+		Machines:     make(map[string]*charm.MachineSpec),
+	}
+
+	if len(model.Applications()) == 0 {
+		return &charm.BundleData{}, errors.NotFoundf("application")
+	}
+	for _, application := range model.Applications() {
+		ut := []string{}
+		for _, unit := range application.Units() {
+			ut = append(ut, unit.Machine().Id())
```
Comment (howbazaar) [inline apiserver/facades/client/bundle/bundle.go]: We don't add placement for subordinate units.

File: apiserver/facades/client/bundle/bundle.go [hunk 17]
```diff
@@ -151,10 +156,111 @@ func (b *BundleAPI) GetChanges(args params.BundleChangesParams) (params.BundleCh
...
+		}
+
+		newApplication := &charm.ApplicationSpec{
+			Charm:            application.CharmURL(),
+			Series:           application.Series(),
+			NumUnits:         len(application.Units()),
```
Comment (howbazaar) [inline apiserver/facades/client/bundle/bundle.go]: Similarly, we don't have a NumUnits for subordinate applications.

File: apiserver/facades/client/bundle/bundle.go (hunk 6 above)
Comment (howbazaar) [reply apiserver/facades/client/bundle/bundle.go]: Yes, there will need to be function to handle the non-existence of various constraint values.

File: apiserver/facades/client/bundle/bundle.go [hunk 18]
```diff
@@ -151,10 +156,111 @@ func (b *BundleAPI) GetChanges(args params.BundleChangesParams) (params.BundleCh
...
+			Series:      machine.Series(),
+		}
+		data.Machines[machine.Id()] = newMachine
+	}
+
+	for _, relation := range model.Relations() {
+		endpointRelation := []string{}
+		for _, endpoint := range relation.Endpoints() {
+			// skipping the 'peer' role which is not of concern in exporting the current model configuration.
+			if endpoint.Role() != "peer" {
```
Comment (howbazaar) [inline apiserver/facades/client/bundle/bundle.go]: You could simplify this loop a little by saying 

```go
if endpoint.Role() == "peer" {
    continue 
}```

File: apiserver/facades/client/bundle/state.go [hunk 19]
```diff
@@ -3,4 +3,15 @@
...
+}
+
+type stateShim struct {
```
Comment (howbazaar) [inline apiserver/facades/client/bundle/state.go]: We don't need this shim.

File: migration/migration.go (hunk 13 above)
Comment (howbazaar) [reply migration/migration.go]: We don't need two interfaces. The `Backend` interface should just have the `ExportPartial` function.

File: apiserver/facades/client/bundle/bundle.go (hunk 3 above)
Comment (anastasiamac) [reply apiserver/facades/client/bundle/bundle.go]: I see! Ok, it makes sense then to marshal it here and return as params.StringResult.

File: apiserver/facades/client/bundle/bundle.go (hunk 15 above)
Comment (vinu2003) [reply apiserver/facades/client/bundle/bundle.go]: @howbazaar , agree that you have mentioned that in other PR that bundles are never accepted in any other format. the discussion started to have similar interface from server like its done in DumpModel command. Agree with your point and I have made it stringResult itself.

File: apiserver/facades/client/bundle/bundle.go (hunk 16 above)
Comment (vinu2003) [reply apiserver/facades/client/bundle/bundle.go]: addressed this.

Comment (vinu2003) [reply apiserver/facades/client/bundle/bundle.go]: description model currently doesnt implement Series(). I will add that.

File: apiserver/facades/client/bundle/bundle.go (hunk 17 above)
Comment (vinu2003) [reply apiserver/facades/client/bundle/bundle.go]: addressed

File: apiserver/facades/client/bundle/bundle.go (hunk 18 above)
Comment (vinu2003) [reply apiserver/facades/client/bundle/bundle.go]: done

File: apiserver/facades/client/bundle/state.go (hunk 19 above)
Comment (vinu2003) [reply apiserver/facades/client/bundle/state.go]: Removed.

File: apiserver/facades/client/bundle/bundle.go (hunk 15 above)
Comment (vinu2003) [reply apiserver/facades/client/bundle/bundle.go]: All this versioning confusion started because i changed from StringResult to BytesResult. Its my fault i didn't take into consideration that gui makes use of this. Apologise.

File: apiserver/facades/client/bundle/bundle.go (hunk 17 above)
Comment (vinu2003) [reply apiserver/facades/client/bundle/bundle.go]: missed these from javascript impl.

File: apiserver/facades/client/bundle/bundle.go (hunk 3 above)
Comment (vinu2003) [reply apiserver/facades/client/bundle/bundle.go]: @howbazaar actually mentioned this in other PR.  that no ther format is accpeted for bundle. the confusion started when i changed it to BytesResult. Apologise. I never considered that gui takes this value.

File: apiserver/facades/client/bundle/bundle.go [hunk 20]
```diff
@@ -5,19 +5,24 @@
...
 
 	"github.com/juju/bundlechanges"
 	"github.com/juju/errors"
 	"github.com/juju/loggo"
 	"gopkg.in/juju/charm.v6"
 	"gopkg.in/juju/names.v2"
+	"gopkg.in/yaml.v2"
 
+	"fmt"
+	"github.com/juju/description"
```
Comment (anastasiamac) [inline apiserver/facades/client/bundle/bundle.go]: move me and the one above

File: apiserver/facades/client/bundle/bundle.go [hunk 21]
```diff
@@ -150,11 +160,149 @@ func (b *BundleAPI) GetChanges(args params.BundleChangesParams) (params.BundleCh
 	return results, nil
 }
 
+var fail = func(failErr error) (params.StringResult, error) {
```
Comment (anastasiamac) [inline apiserver/facades/client/bundle/bundle.go]: Please move this declaration to inside the method where it is relevant. Only ExportBundle uses it.

File: apiserver/facades/client/bundle/bundle.go [hunk 22]
```diff
@@ -150,11 +160,149 @@ func (b *BundleAPI) GetChanges(args params.BundleChangesParams) (params.BundleCh
...
 func (b *BundleAPI) ExportBundle() (params.StringResult, error) {
-	return params.StringResult{}, nil
+	if err := b.checkCanRead(); err != nil {
+		return fail(err)
+	}
+
+	var exportConfig state.ExportConfig
+
+	b.backend.SetExportconfig(exportConfig)
+	logger.Criticalf("XXXXXXXXXXXXXXXXX...........SkipStatusHistory: %v ", exportConfig.SkipStatusHistory)
```
Comment (anastasiamac) [inline apiserver/facades/client/bundle/bundle.go]: This does not belong in production code ;)

File: apiserver/facades/client/bundle/bundle.go [hunk 23]
```diff
@@ -67,10 +72,15 @@ func newFacade(ctx facade.Context) (*BundleAPI, error) {
...
+	model, err := st.Model()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
 	return NewBundleAPI(
-		st,
+		NewStateShim(st),
 		authorizer,
-		names.NewModelTag(st.ModelUUID()),
```
Comment (anastasiamac) [inline apiserver/facades/client/bundle/bundle.go]: Why did you change this? Previous implementation was laconic and you did not need to go to .Model().

File: apiserver/facades/client/bundle/bundle.go [hunk 24]
```diff
@@ -150,11 +160,149 @@ func (b *BundleAPI) GetChanges(args params.BundleChangesParams) (params.BundleCh
...
 // ExportBundle exports the current model configuration as bundle.
 func (b *BundleAPI) ExportBundle() (params.StringResult, error) {
-	return params.StringResult{}, nil
+	if err := b.checkCanRead(); err != nil {
+		return fail(err)
+	}
+
+	var exportConfig state.ExportConfig
+
+	b.backend.SetExportconfig(exportConfig)
```
Comment (anastasiamac) [inline apiserver/facades/client/bundle/bundle.go]: Why do u need to it this way? What we discussed this morning would have looked something like https://pastebin.ubuntu.com/p/skHsmffKgb/

This way, you next call becomes

```
b.backend.ExportPartial()
```

File: apiserver/facades/client/bundle/bundle.go [hunk 25]
```diff
@@ -150,11 +160,149 @@ func (b *BundleAPI) GetChanges(args params.BundleChangesParams) (params.BundleCh
...
+		Result: string(bytes),
+	}, nil
 }
 
 // Mask the new method from V1 API.
 // ExportBundle is not in V1 API.
 func (u *APIv1) ExportBundle() (_, _ struct{}) { return }
+
+// FillBundle fills the bundledata datastructure required for the exportBundle.
+func (b *BundleAPI) FillBundleData(model description.Model) (*charm.BundleData, error) {
```
Comment (anastasiamac) [inline apiserver/facades/client/bundle/bundle.go]: Please do not export this. This should be fillBundleData...

File: apiserver/facades/client/bundle/bundle.go [hunk 26]
```diff
@@ -150,11 +160,149 @@ func (b *BundleAPI) GetChanges(args params.BundleChangesParams) (params.BundleCh
...
+	cfg := model.Config()
+	var defaultSeries string
```
Comment (anastasiamac) [inline apiserver/facades/client/bundle/bundle.go]: This may leave defaultSeries uninitialised...

Something like this 
defaultSeries, ok := cfg["default-series"]
if !ok {
 // should never happen, but just in case it does
// let's set it to latest lts
}

and the for-loop becomes redundant... no need to iterate over the whole of config if we are only after one value.

File: apiserver/facades/client/bundle/bundle.go [hunk 27]
```diff
@@ -150,11 +160,149 @@ func (b *BundleAPI) GetChanges(args params.BundleChangesParams) (params.BundleCh
...
+
+	data := &charm.BundleData{
+		Series:       defaultSeries,
+		Applications: make(map[string]*charm.ApplicationSpec),
+		Relations:    [][]string{},
+		Machines:     make(map[string]*charm.MachineSpec),
+	}
+
+	if len(model.Applications()) == 0 {
+		return &charm.BundleData{}, errors.Errorf("nothing to export as there is no application found.")
```
Comment (anastasiamac) [inline apiserver/facades/client/bundle/bundle.go]: Maybe "... there are no applications..."

File: apiserver/facades/client/bundle/bundle.go [hunk 28]
```diff
@@ -150,11 +160,149 @@ func (b *BundleAPI) GetChanges(args params.BundleChangesParams) (params.BundleCh
...
+			}
+		}
+
+		data.Applications[application.Name()] = newApplication
+	}
+
+	for _, machine := range model.Machines() {
+		result := b.constraints(machine.Constraints())
+
+		constraints := []string{"arch=" + *result.Arch,
```
Comment (anastasiamac) [inline apiserver/facades/client/bundle/bundle.go]: We do not want to create all this if thher are not specified...

I think you want your b.constraints(...) to return this collection populating only the parameters that were specified...

File: apiserver/facades/client/bundle/bundle_test.go [hunk 29]
```diff
@@ -214,3 +215,12 @@ func (s *bundleSuite) TestGetChangesBundleEndpointBindingsSuccess(c *gc.C) {
 		}
 	}
 }
+
+func (s *bundleSuite) TestExportBundleFailNoApplication(c *gc.C) {
+	var cfg state.ExportConfig
+	model, err := s.st.Exportpartial(cfg)
```
Comment (anastasiamac) [inline apiserver/facades/client/bundle/bundle_test.go]: This is not testing the facade though...

File: apiserver/facades/client/bundle/bundle.go [hunk 30]
```diff
@@ -5,19 +5,24 @@
...
 
 	"github.com/juju/bundlechanges"
 	"github.com/juju/errors"
 	"github.com/juju/loggo"
+	"github.com/juju/os/series"
 	"gopkg.in/juju/charm.v6"
 	"gopkg.in/juju/names.v2"
+	"gopkg.in/yaml.v2"
 
+	"github.com/juju/description"
```
[the remark this answers is not in the document]
Comment (anastasiamac) [reply apiserver/facades/client/bundle/bundle.go]: This import is not in the right group.

File: apiserver/facades/client/bundle/bundle.go [hunk 31]
```diff
@@ -152,9 +157,142 @@ func (b *BundleAPI) GetChanges(args params.BundleChangesParams) (params.BundleCh
 
 // ExportBundle exports the current model configuration as bundle.
 func (b *BundleAPI) ExportBundle() (params.StringResult, error) {
-	return params.StringResult{}, nil
+	var fail = func(failErr error) (params.StringResult, error) {
```
Comment (anastasiamac) [inline apiserver/facades/client/bundle/bundle.go]: This could be just a  local declaration
```
 fail := ...
```

rather than a 
```
var fail = 
```

File: apiserver/facades/client/bundle/bundle.go [hunk 32]
```diff
@@ -152,9 +157,142 @@ func (b *BundleAPI) GetChanges(args params.BundleChangesParams) (params.BundleCh
...
+	return params.StringResult{
+		Result: string(bytes),
+	}, nil
 }
 
 // Mask the new method from V1 API.
 // ExportBundle is not in V1 API.
 func (u *APIv1) ExportBundle() (_, _ struct{}) { return }
+
+// FillBundle fills the bundledata datastructure required for the exportBundle.
```
Comment (anastasiamac) [inline apiserver/facades/client/bundle/bundle.go]: Comment needs to match func name and is not necessary for unexported (starting with lower-case) functions unless they are complex. If you decide to keep the comment, however, 'data structure' might be missing a space?..

File: apiserver/facades/client/bundle/bundle.go [hunk 33]
```diff
@@ -152,9 +157,142 @@ func (b *BundleAPI) GetChanges(args params.BundleChangesParams) (params.BundleCh
...
+func (b *BundleAPI) fillBundleData(model description.Model) (*charm.BundleData, error) {
+	// get the default-series from model config.
```
Comment (anastasiamac) [inline apiserver/facades/client/bundle/bundle.go]: The comment is not needed as the code is self-explanatory.

File: apiserver/facades/client/bundle/bundle.go [hunk 34]
```diff
@@ -152,9 +157,142 @@ func (b *BundleAPI) GetChanges(args params.BundleChangesParams) (params.BundleCh
...
+				Annotations:      application.Annotations(),
+				EndpointBindings: application.EndpointBindings(),
+			}
+		}
+
+		data.Applications[application.Name()] = newApplication
+	}
+
+	for _, machine := range model.Machines() {
+		var constraints string
```
Comment (anastasiamac) [inline apiserver/facades/client/bundle/bundle.go]: This whole paragraph can be simplified to 
```
		newMachine := &charm.MachineSpec{
			Annotations: machine.Annotations(),
			Series:      machine.Series(),
		}
		if result := b.constraints(machine.Constraints()); len(result) != 0 {
			newMachine.Constraints = strings.Join(result, " ")
		}
```

File: apiserver/facades/client/bundle/bundle.go [hunk 35]
```diff
@@ -152,9 +157,142 @@ func (b *BundleAPI) GetChanges(args params.BundleChangesParams) (params.BundleCh
...
+		}
+	}
+	data.Relations = append(data.Relations, endpointRelation)
+
+	return data, nil
+}
+
+func (b *BundleAPI) constraints(cons description.Constraints) []string {
+	if cons == nil {
+		return []string{}
```
Comment (anastasiamac) [inline apiserver/facades/client/bundle/bundle.go]: Why not nil?

File: apiserver/facades/client/bundle/bundle.go [hunk 36]
```diff
@@ -152,9 +157,142 @@ func (b *BundleAPI) GetChanges(args params.BundleChangesParams) (params.BundleCh
...
+
+	data := &charm.BundleData{
+		Series:       defaultSeries,
+		Applications: make(map[string]*charm.ApplicationSpec),
+		Relations:    [][]string{},
+		Machines:     make(map[string]*charm.MachineSpec),
+	}
+
+	if len(model.Applications()) == 0 {
+		return &charm.BundleData{}, errors.Errorf("nothing to export as there are no applications.")
```
Comment (anastasiamac) [inline apiserver/facades/client/bundle/bundle.go]: Error messages should not contain a full stop as other layers may append more information to the message.

It's kind of counter-intuitive until you see the trace coming from cmd layer :)

File: apiserver/facades/client/bundle/bundle_test.go [hunk 37]
```diff
@@ -214,3 +217,447 @@ func (s *bundleSuite) TestGetChangesBundleEndpointBindingsSuccess(c *gc.C) {
...
+			"uuid": "some-uuid",
+		},
+		CloudRegion: "some-region"})
+	s.st.model.SetStatus(description.StatusArgs{Value: "available"})
+
+	result, err := s.facade.ExportBundle()
+	c.Assert(err, gc.NotNil)
+	c.Assert(result, gc.Equals, params.StringResult{})
+	c.Check(err, gc.ErrorMatches, "nothing to export as there are no applications.")
+	s.st.CheckCallNames(c, "ExportPartial")
```
Comment (anastasiamac) [inline apiserver/facades/client/bundle/bundle_test.go]: This is an unnecessary check here and everywhere else, as the following line checks this and more ;D

**Comment by vinu2003** [comment]
!!build!!

**Comment by vinu2003** [comment]
$$merge$$
